File: library/src/main/java/com/snail/commons/helper/ApkInstallHelper.java
package com.snail.commons.helper;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.provider.Settings;
import androidx.annotation.NonNull;
import com.snail.commons.util.FileUtils;

import java.io.File;
import java.util.Objects;

/**
 * apk安装帮助类
 * <p>
 * date: 2019/8/6 14:03
 * author: zengfansheng
 */
public class ApkInstallHelper {
    private static final int REQUEST_CODE = 3984;
    private final Activity activity;
    private File apkFile;

    public ApkInstallHelper(@NonNull Activity activity) {
        this.activity = activity;
    }

    /**
     * 如果是Android8.0以上需要在Activity中的onActivityResult调用此方法
     */
    public void onActivityResult(int requestCode) {
        if (requestCode == REQUEST_CODE && Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            if (activity.getPackageManager().canRequestPackageInstalls()) {
                install();
            }
        }
    }

    /**
     * 安装apk
     */
    public void install(@NonNull File apkFile) {
        Objects.requireNonNull(apkFile, "apkFile is null");
        this.apkFile = apkFile;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && !activity.getPackageManager().canRequestPackageInstalls()) {
            Uri uri = Uri.parse("package:" + activity.getPackageName());
            Intent intent = new Intent(Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES, uri);
            activity.startActivityForResult(intent, REQUEST_CODE);
        } else {
            install();
        }
    }

    private void install() {
        if (apkFile.exists()) {
            Intent intent = new Intent(Intent.ACTION_VIEW);
            FileUtils.setIntentDataAndType(apkFile, activity, intent, "application/vnd.android.package-archive", false);
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            activity.startActivity(intent);
        }
    }
}

File: library/src/main/java/com/snail/commons/entity/ApkInstallHelper.kt
package com.snail.commons.entity

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.provider.Settings
import com.snail.commons.utils.setIntentDataAndType
import java.io.File

/**
 * apk安装帮助类
 *
 * date: 2019/5/8 10:42
 * author: zengfansheng
 */
class ApkInstallHelper(private val activity: Activity) {
    private var apkFile: File? = null
    /**
     * 如果是Android8.0以上需要在Activity中的onActivityResult调用此方法
     */
    fun onActivityResult(requestCode: Int) {
        if (requestCode == REQUEST_CODE && Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            if (activity.packageManager.canRequestPackageInstalls()) {
                install()
            }
        }
    }

    /**
     * 只有下载的是apk文件时
     */
    fun install(apkFile: File) {
        this.apkFile = apkFile
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && !activity.packageManager.canRequestPackageInstalls()) {
            val intent = Intent(Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES, Uri.parse("package:" + activity.packageName))
            activity.startActivityForResult(intent, REQUEST_CODE)
        } else {
            install()
        }
    }

    private fun install() {
        if (apkFile?.exists() == true) {
            val intent = Intent(Intent.ACTION_VIEW)
            apkFile!!.setIntentDataAndType(activity, intent, "application/vnd.android.package-archive", false)
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            activity.startActivity(intent)
        }
    }

    companion object {
        private const val REQUEST_CODE = 3984
    }
}